master/internal/task/allocation.go

```diff
@@ -215,6 +218,18 @@ func (a *Allocation) Receive(ctx *actor.Context) error {
 		a.Terminate(ctx, "allocation resource pool changed", false)
 	case actor.PostStop:
 		a.Cleanup(ctx)
+		// if this is a restore allocation req and a.restored is set
+		// then the ports have been added to the port registry successfully.
+		// This is to avoid the race condition of a failed restored
+		// allocation releasing another allocation's port(s).
+		// a.restorePorts is set to true right after ports are restored.
+		// This variable ensures to release ports even if there's a failure after restoring ports.
+		// The a.req.Restore == a.restored  (XNOR) condition will be true even if it's a new allocation.
+		if a.req.Restore == a.restored || a.restorePorts { // XNOR
```

Comment:
probably a lot easier to follow if we change `a.restorePorts` to `a.portsRegistered` and flip it after all the alls to `RestorePort` and `GetPort` are finished